# ViSpace Pipeline — Interactive Notebook

End-to-end whole-slide image (WSI) analysis for H&E pathology slides:  
tile extraction → semantic segmentation → spatial feature extraction.

---

## Pipeline overview

```
WSI (.svs/.tif)
  │
  ├─ Stage 1 · Tessellation          tile the slide with Mussel
  ├─ Stage 2 · Segmentation          run segmentation model per tile
  ├─ Stage 3 · Stitching             assemble tiles → GeoJSON + PNG
  │
  ├─ Stage 4 · Tumor ROI Overlay     cluster tumour tiles → ROI boxes
  ├─ Stage 5 · Cluster TSR/sTILs     score TSR & TILs per cluster polygon
  │
  ├─ Stage 6 · Immune Proximity      TIL distance/phenotype features
  ├─ Stage 7 · Necrosis Proximity    necrosis distance/phenotype features
  └─ Stage 8 · Tumour Morphology     island shape, fragmentation, NND
```

Every stage writes its outputs to a predictable subdirectory under `cfg.OUT_DIR/<slide_name>/`.
If an output already exists the stage is **automatically skipped**, so interrupted runs resume cleanly.

---

**How to use this notebook**

1. Make sure you go through all the prechecks in **Precheck 1-3**.
2. Run cells after that top-to-bottom for a guided walkthrough, *or*
3. Jump straight to **Cell 4** to run the full pipeline in one call.
4. Use later cells to inspect outputs interactively after the run.

## Precheck 1 — Hugging Face Login

Make sure you have the hugging face login set

In [1]:
# Cell 1 — HF auth
import os
from huggingface_hub import whoami

# ✏ Paste your HF access token here (Read role is sufficient)
HF_ACCESS_TOKEN = "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"

# Verify the token is valid and print your username
user = whoami(token=HF_ACCESS_TOKEN)
print(f"Authenticated as: {user['name']}")

# Expose the token to all HF libraries for the rest of the session
os.environ["HF_TOKEN"] = HF_ACCESS_TOKEN
print("HF_TOKEN environment variable set.")

Authenticated as: himangi2003
HF_TOKEN environment variable set.


## Precheck 2 — Set Pipeline Configuration

Set `WSI_PATH` to your slide file and `OUT_DIR` to where results should be written. 

All other settings have sensible defaults adjust only what you need. 
For reference chec the `config.py` file and Pipeline manual document in the folder 

| Parameter | Meaning | Default |
|-----------|---------|---------|
| `WSI_PATH` | Path to the `.svs` / `.tif` slide | *(required)* |
| `OUT_DIR` | Root output directory | *(required)* |
| `CHECKPOINT` | Path to the ViSpace segmentation model checkpoint | *(required)* |


In [2]:
# Cell 2 — build config
from config import PipelineConfig

cfg = PipelineConfig(
    WSI_PATH   = "/lab/deasylab3/Jung/Data/Shared_Data/TCGA_TNBC/histology/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0.tif",
    CHECKPOINT = "weights_frozen_encoder/TNBC_best.pt",
    OUT_DIR    = "vipsegd_output",
    BATCH_SIZE = 96
    
)
cfg


PipelineConfig(OUT_DIR='vipsegd_output', CHECKPOINT='weights_frozen_encoder/TNBC_best.pt', WSI_PATH='/lab/deasylab3/Jung/Data/Shared_Data/TCGA_TNBC/histology/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0.tif', MPP=0.25, PATCH_SIZE=224, WORKERS=4, SEGMENT_THRESH=20, THUMBNAIL_SIZE=(1024, 1024), VIRCHOW2_PATH=None, DEVICE='cuda', BATCH_SIZE=96, WHITE_THRESH=220, STEP_X=444, STEP_Y=444, MAX_PX=4096, ALPHA=0.6, THUMB_MAX=2048, ROI_SIZE_UM=200.0, ROI_MIN_TUMOR_FRAC=0.2, ROI_MAX_NECROSIS=0.5, ROI_MIN_CLUSTER_TILES=3, ROI_CONNECTIVITY=8, ROI_MERGE_GAP_UM=60.0, ROI_THUMB_WIDTH=1800, ROI_COLOR_BY_CLUSTER=True, TILS_DENOMINATOR='salgado', CLUSTER_BUFFER_UM=300.0, CLUSTER_MIN_ROI_BOXES=1, CLUSTER_MIN_POLYGON_AREA_PX2=1.0, CLUSTER_MAX_AREA_QUANTILE=1.0, CLUSTER_MAX_AREA_MAD_Z=0.0, CLUSTER_MIN_TISSUE_FRACTION=0.1, CLUSTER_MIN_TSR_DENOM_PX2=5000.0, CLUSTER_MIN_TILS_DENOM_PX2=5000.0, CLUSTER_NO_OVERLAY=False, IMMUNE_PROXIMITY_THRESHOLDS_UM=[20, 50, 100, 200], IMMUNE_CONTACT_TOLERANCE_U

## Precheck 3 — Environment and GPU check

Verify that all required packages are importable before starting and also check GPU compatibility and compute
If anything is missing the `pip install` line will install it.

In [3]:
from environment_check import run_environment_check

if not run_environment_check(cfg):
    raise SystemExit("Fix the environment issues above before running.")




  Python version
  Running     : Python 3.11.15
  [OK]   Python 3.11 meets minimum 3.10

  Required packages
  [OK]   mussel             (mussel-pathology) — 1.4.3
  [OK]   torch              (torch) — 2.5.1+cu121
  [OK]   torchvision        (torchvision) — 0.20.1+cu121
  [OK]   timm               (timm) — 1.0.17
  [OK]   huggingface_hub    (huggingface_hub) — 0.33.4
  [OK]   numpy              (numpy) — 2.2.6
  [OK]   pandas             (pandas) — 2.3.1
  [OK]   cv2                (opencv-python-headless) — 4.12.0.88
  [OK]   PIL                (pillow) — 11.3.0
  [OK]   tqdm               (tqdm) — 4.67.1
  [OK]   shapely            (shapely) — 2.1.1
  [OK]   geopandas          (geopandas) — 1.1.1
  [OK]   scipy              (scipy) — 1.16.0
  [OK]   matplotlib         (matplotlib) — 3.10.3
  [OK]   ml_dtypes          (ml_dtypes) — 0.5.4
  [OK]   open_clip          (open_clip_torch) — 3.3.0
  [OK]   omegaconf          (omegaconf) — 2.3.0
  [OK]   h5py               (h5py) — 3.14.0
  

---
## Stage 1: Tessellation

**What happens:** Mussel tiles the WSI at 224 × 224 pixels.
An Otsu tissue mask is computed to skip glass/background tiles.
Outputs a directory of PNG patches and an HDF5 index.

**Key outputs:**
- `tessellation/patches/<slide>_<wx>_<wy>.png` — one file per tissue tile
- `tessellation/<slide>.h5` — coordinate index (sentinel file for this stage)
- `tessellation/thumbnail.png` — low-resolution slide overview
- `tessellation/mask.png` — binary tissue mask

**Config knobs:** `SEGMENT_THRESH`, `THUMBNAIL_SIZE`, `WORKERS`

In [4]:
# — tessellate
from tessellate import run_tessellation

tess_dir = run_tessellation(cfg.WSI_PATH, cfg)
tess_dir


  ViSpace Tessellation
  Slide          : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0.tif
  Slide path     : /lab/deasylab3/Jung/Data/Shared_Data/TCGA_TNBC/histology/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0.tif
  Mussel version : 1.4.3
  Patch size     : 224
  Workers        : 4
  Segment thresh : 20
  Thumbnail size : (1024, 1024)
  Output         : /cluster/home/srivash/ViP-SegD/VipSegD_v1/vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/tessellation

  Running Mussel tessellation. Large slides may take some time...

  Tessellation completed successfully.
  Patch images : 7,030
  HDF5 output  : /cluster/home/srivash/ViP-SegD/VipSegD_v1/vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/tessellation/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0.h5
  Output folder: /cluster/home/srivash/ViP-SegD/VipSegD_v1/vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/t

'/cluster/home/srivash/ViP-SegD/VipSegD_v1/vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/tessellation'

## Stage 2: Segmentation

**What happens:** Loads the ViSpace segmenter model from `cfg.CHECKPOINT`.  Each tile is passed through the model to produce a per-pixel class map saved as a `.npy` file.  Near-white background pixels are masked to `SEG_IGNORE`.

**Classes:** Tumour · Stroma · Inflammatory · Necrosis · Others

**Key outputs:**
- `segmentation/<tile>_seg.npy` — per-tile int32 class map *(temporary, deleted by Stage 3)*
- `segmentation/manifest.csv` — `tile, wx, wy, npy_path` for every predicted tile
  *(sentinel file for this stage)*

**Config knobs:** `CHECKPOINT` for the segmentation checkpoint you need to use.

> **Hardware note:** this stage runs on GPU if `torch.cuda.is_available()`, otherwise CPU.
> Expect ~3–10 min for a typical TCGA slide on 4 GPU.


In [5]:
# segment
from segmenter import run_segmentation

manifest_csv = run_segmentation(cfg.WSI_PATH, cfg)
manifest_csv


  Segmentation
  Slide      : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  Tiles      : 7,030
  Batch size : 96
  Device     : cuda
  Checkpoint : TNBC_best.pt
  Output     : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation

  Loading checkpoint: TNBC_best.pt
  Loading Virchow2 from: hf-hub:paige-ai/Virchow2
  Virchow2: 0.63B params | frozen | cuda
  Decoder params: 9.5M
  AMP dtype  : torch.float16


Running segmentation: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7030/7030 [01:23<00:00, 84.26tile/s]



  Manifest saved → vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/manifest.csv
  Done. 7,030 tiles segmented → vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation


'vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/manifest.csv'

## Stage 3: Stitching

**What happens:** Reads every per-tile `.npy` class map from the manifest, resizes
each to fill its grid slot (eliminating checkerboard gaps), and places it onto a
large integer canvas in WSI level-0 pixel coordinates.

The stitched canvas is then vectorised into GeoJSON polygons — one feature per
connected region, with `classification.name` set to the tissue class.  This
GeoJSON is the input for all downstream spatial-analysis stages.

Finally, all `.npy` scratch files are **automatically deleted** once both final
outputs exist.

**Key outputs:**
- `segmentation/segmentation_all_classes.geojson` ← **primary input for stages 4–8**
- `segmentation/<slide>_segmentation.png` — colour overlay with legend

**Config knobs:** `MPP` (used for axis step inference fallback)


In [6]:
# stitch + GeoJSON
from stitch import run_stitching

stitch_results = run_stitching(cfg.WSI_PATH, cfg)
stitch_results


  Stitching + GeoJSON
  Slide      : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  Manifest   : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/manifest.csv
  Output     : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation

  Stitching
  Slide      : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  Tiles      : 7,030
  Canvas     : 49062x25086 px  origin=(3265,9731)
  Step       : 222x222  (auto-inferred)


Stitching segmentation tiles: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7030/7030 [00:04<00:00, 1506.10tile/s]


  Saved npy  : stitched_seg.npy
  Saved png  : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0_segmentation.png

  Class distribution:
    Tumour               946,694 px  (0.3%)  
    Stroma               169,229,982 px  (52.2%)  ████████████████████
    Inflammatory         3,579,466 px  (1.1%)  
    Necrosis             99,536,108 px  (30.7%)  ████████████
    Others               50,811,960 px  (15.7%)  ██████

  GeoJSON extraction
  Canvas   : 49062x25086 px  origin=(3265,9731)
    Tumour               946,694 px  (0.3%)
    Stroma               169,229,982 px  (52.2%)
    Inflammatory         3,579,466 px  (1.1%)
    Necrosis             99,536,108 px  (30.7%)
    Others               50,811,960 px  (15.7%)


Writing GeoJSON:  20%|██████████████████████████████████████                                                                                                                                                        | 1/5 [00:04<00:16,  4.13s/class]

  Tumour                 257 polygons  ~0.53 MB


Writing GeoJSON:  40%|████████████████████████████████████████████████████████████████████████████                                                                                                                  | 2/5 [00:11<00:18,  6.08s/class]

  Stroma               5,714 polygons  ~22.06 MB


Writing GeoJSON:  60%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████                                                                            | 3/5 [00:15<00:10,  5.29s/class]

  Inflammatory         1,927 polygons  ~3.37 MB


Writing GeoJSON:  80%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████                                      | 4/5 [00:21<00:05,  5.40s/class]

  Necrosis             3,741 polygons  ~13.17 MB


Writing GeoJSON: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:28<00:00,  5.64s/class]

  Others               5,681 polygons  ~17.69 MB



  Combined : 17,320 polygons  51.21 MB  → segmentation_all_classes.geojson


Cleaning temporary files: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7031/7031 [00:01<00:00, 5788.52file/s]

  Cleaned up : 7,031 .npy files from segmentation/

  Done.
  Colour PNG   : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0_segmentation.png
  Combined GJ  : segmentation_all_classes.geojson
  .npy removed : 7,031


{'slide_name': 'TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0',
 'manifest_csv': 'vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/manifest.csv',
 'stitched_png': 'vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0_segmentation.png',
 'geojson': {'combined': 'vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/segmentation_all_classes.geojson'},
 'meta': {'manifest_csv': 'vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/manifest.csv',
  'stitched_seg_npy': 'vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/stitched_seg.npy',
  'stitched_seg_png': 'vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0_segmentation.png',
  'slid

## Stage 4: Tumor ROI Overlay

**What happens:** Reads the tile manifest and filters for tumour-rich tiles
(`frac_Tumour ≥ cfg.ROI_MIN_TUMOR_FRAC`), then runs a 5-step pipeline:

1. **Filter** tumour tiles by minimum tumour fraction
2. **Cluster** tiles into spatially-connected clumps via BFS on the tile grid
3. **Merge** nearby clumps within `cfg.ROI_MERGE_GAP_UM` of each other (union-find)
4. **Box** each clump with non-overlapping square ROI boxes of `cfg.ROI_SIZE_UM` µm
5. **Overlay** boxes on a pseudo-thumbnail and optionally the real WSI thumbnail

**Key outputs:**
- `tumor_roi_overlay/tumor_roi_boxes.csv` ← **input for Stage 5**
- `tumor_roi_overlay/tumor_roi_boxes_pseudo_thumbnail.png`
- `tumor_roi_overlay/tumor_roi_boxes_wsi_thumbnail.png` *(if WSI file exists on disk)*

**Config knobs:** `MPP`, `ROI_SIZE_UM`, `ROI_MIN_TUMOR_FRAC`, `ROI_MAX_NECROSIS`,
`ROI_MIN_CLUSTER_TILES`, `ROI_CONNECTIVITY`, `ROI_MERGE_GAP_UM`

> **Why ROI boxes?** The GeoJSON segmentation covers the entire scanned area including
> stroma-only and background regions. ROI boxes focus downstream scoring polygons
> on tissue regions that actually contain tumour.


In [ ]:
# tumor ROI clustering + overlay
from tumor_roi_overlay import run_tumor_roi_overlay

roi_results = run_tumor_roi_overlay(cfg.WSI_PATH, cfg)
roi_results

In [ ]:
# Cell 11 (optional) — inspect a result inline, e.g. the ROI overlay image
from PIL import Image
Image.open(roi_results["pseudo_png"])

## Stage 5: Cluster TSR / sTILs Scoring

**What happens:** Dissolves the ROI boxes for each tumour cluster into a single
convex scoring polygon, buffers it by `cfg.CLUSTER_BUFFER_UM` µm to capture
peritumoural stroma, then clips and non-overlapping-assigns polygons across clusters.

For each scoring polygon, all GeoJSON segmentation polygons that intersect it are
clipped and summed per class.  Background (unsegmented pixels) is tracked separately
and **never** enters any denominator.

**Metrics computed:**

| Metric | Formula | Denominator |
|--------|---------|-------------|
| TSR | Stroma / (Tumour + Stroma) | T + S only |
| sTILs (Salgado) | Inflammatory / Stroma × 100 | Stroma |
| sTILs (stromal) | Inflammatory / (Stroma + Inflam) × 100 | S + I |
| sTILs (tissue) | Inflammatory / Viable × 100 | T+S+I+O |

**Key outputs:**
- `cluster_tils_tsr_score/cluster_scoring_polygons.geojson` ← **input for stages 6–8**
- `cluster_tils_tsr_score/tils_tsr_by_cluster.csv`
- `cluster_tils_tsr_score/tils_tsr_wsi_summary.csv`
- `cluster_tils_tsr_score/cluster_tils_tsr_overlay.png`


In [ ]:
# cluster TSR / sTILs scoring
from cluster_tils_tsr_scoring import run_cluster_tils_tsr_score

tsr_results = run_cluster_tils_tsr_score(cfg.WSI_PATH, cfg)
tsr_results["wsi_summary"]

##  Stage 6: Immune Proximity Features

**What happens:** For each cluster scoring polygon, all Inflammatory-class polygons
are clipped to the cluster boundary.  For each resulting inflammatory component,
its `representative_point()` (guaranteed inside the polygon) is located relative
to the tumour boundary:

- **Negative distance** → inside tumour (intratumoral TIL)
- **Positive distance** → outside tumour (extratumoral TIL)

**Immune spatial phenotype** (Galon & Bruni 2020 classification hierarchy):

| Phenotype | Criterion |
|-----------|-----------|
| `immune-desert` | total TIL area < 1,000 µm² |
| `immune-penetrated` | ≥ 30% intratumoral |
| `margin-localized` | ≥ 50% within 50 µm of boundary |
| `immune-excluded` | median extratumoral distance ≥ 100 µm |
| `peritumoral` | present, close, but not penetrating |

**Key outputs:**
- `immune_proximity/immune_proximity_by_cluster.csv`
- `immune_proximity/immune_proximity_wsi_summary.csv`
- `immune_proximity/immune_proximity_plot.png`


In [ ]:
# immune / TIL proximity features
from immune_proximity_features import run_immune_proximity_features

immune_results = run_immune_proximity_features(cfg.WSI_PATH, cfg)
immune_results

## Stage 7: Necrosis Proximity Features

**What happens:** The same signed-distance approach as Stage 6, but applied to
Necrosis-class components.  Additionally computes nearest-distance to both
Inflammatory and Stroma polygons within each cluster (TNF-proxy / stromal
remodelling proxy).

**Necrosis spatial phenotype** (classification hierarchy):

| Phenotype | Criterion |
|-----------|-----------|
| `necrosis-absent` | total area < 500 µm² |
| `tumour-central` | intratumoral fraction ≥ 50% |
| `peritumoural` | ≥ 60% within 100 µm of tumour boundary |
| `immune-adjacent` | ≥ 20% of necrosis near Inflammatory tissue |
| `stromal-distant` | present but spatially decoupled |

**Extra features vs immune proximity:**
- `necrosis_immune_coupling_index` — fraction of necrosis within `NECROSIS_IMMUNE_COUPLING_THRESHOLD_UM` of inflammatory tissue
- `necrosis_stroma_min_dist_aw_median_um` — distance to nearest stroma
- `necrosis_to_tumor_ratio` — clinically used ratio

**Key outputs:**
- `necrosis_proximity/necrosis_proximity_by_cluster.csv`
- `necrosis_proximity/necrosis_proximity_wsi_summary.csv`
- `necrosis_proximity/necrosis_distance_figure.png`
- `necrosis_proximity/necrosis_tissue_context_figure.png`

In [ ]:
# necrosis proximity features
from necrosis_proximity_features import run_necrosis_proximity_features

necrosis_results = run_necrosis_proximity_features(cfg.WSI_PATH, cfg)
necrosis_results

## Stage 8: Tumour Morphology Features

**What happens:** Tumour-class polygons from the segmentation GeoJSON are clipped
to each cluster scoring polygon, then dissolved into a single multi-island geometry.
Each connected island is measured individually and aggregated with area-weighting
(large nests count more than tiny fragments).

> **Important:** always set `cfg.MORPHOLOGY_MIN_ISLAND_AREA_UM2 ≥ 1000` to filter
> single-cell pixel-segmentation fragments before computing shape metrics.
> Report this threshold alongside any shape metric.

**Feature groups:**

| Group | Features |
|-------|---------|
| Tumour burden | `tumor_area_mm2`, `tumor_fraction_of_cluster`, `tumor_perimeter_um` |
| Shape (area-weighted) | `tumor_compactness_mean`, `tumor_solidity_mean`, `tumor_elongation_mean` |
| Fragmentation | `tumor_n_islands`, `tumor_largest_patch_index`, `tumor_fragmentation_index`, `tumor_effective_patches` |
| Size distribution | `tumor_island_area_mean/median/p90/max_um2`, `tumor_island_area_cv` |
| Spatial spread | `tumor_island_nnd_median_um`, `tumor_spread_frac`, `tumor_convex_hull_fill` |
| Size tiers | `tumor_n_islands_fragment/micro/small/large` |

**Key outputs:**
- `tumor_morphology/tumor_core_features_by_cluster.csv`
- `tumor_morphology/tumor_core_wsi_summary.csv`
- `tumor_morphology/tumor_island_qc.csv` *(when `cfg.MORPHOLOGY_SAVE_ISLAND_QC = True`)*


In [ ]:
# tumor morphology features
from tumor_morphology_features import run_tumor_morphology_features

morphology_results = run_tumor_morphology_features(cfg.WSI_PATH, cfg)
morphology_results


##  Run the full pipeline in one go

One call to `run_vispace` executes all eight stages in dependency order.

**Resumable** — stages whose primary output already exists on disk are skipped automatically.
Re-run this cell after a crash and it will pick up from where it left off.

**Selective re-run** — pass `force_stages={...}` to re-execute specific stages:
```python
results = run_vispace (WSI_PATH, cfg, force_stages={'segmentation', 'stitching'})
```

**Subset run** — pass `stages={...}` to run only certain stages
(their prerequisites must already be on disk):
```python
results = run_vispace (WSI_PATH, cfg, stages={'immune_proximity', 'tumor_morphology'})
```

In [4]:
# run all 8 stages end-to-end
from run_vispace import run_vispace

results = run_vispace(cfg.WSI_PATH, cfg)
results["success"]

failed = {name: info for name, info in results['stages'].items()
          if info.get('status') == 'failed'}

if not failed:
    print("No stages failed in the last run.")
else:
    for stage_name, info in failed.items():
        print("=" * 60)
        print(f"FAILED STAGE: {stage_name}")
        print("=" * 60)
        print(info["error"])



############################################################
  ViSPACE pipeline
  Slide : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  Stages: tessellation, segmentation, stitching, tumor_roi_overlay, cluster_tils_tsr_score, immune_proximity, necrosis_proximity, tumor_morphology
############################################################

  ▶  [tessellation] starting …

  ViSpace Tessellation
  Slide          : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0.tif
  Slide path     : /lab/deasylab3/Jung/Data/Shared_Data/TCGA_TNBC/histology/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0.tif
  Mussel version : 1.4.3
  Patch size     : 224
  Workers        : 4
  Segment thresh : 20
  Thumbnail size : (1024, 1024)
  Output         : /cluster/home/srivash/ViP-SegD/VipSegD_v1/vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/tessellation

  Running Mussel tessellation. Large slides may take some time...

  Tessellation com

Running segmentation: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7030/7030 [01:25<00:00, 82.01tile/s]



  Manifest saved → vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/manifest.csv
  Done. 7,030 tiles segmented → vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation
  ✓  [segmentation] done in 104.5s

  ▶  [stitching] starting …

  Stitching + GeoJSON
  Slide      : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  Manifest   : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/manifest.csv
  Output     : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation

  Stitching
  Slide      : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  Tiles      : 7,030
  Canvas     : 49062x25086 px  origin=(3265,9731)
  Step       : 222x222  (auto-inferred)


Stitching segmentation tiles: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7030/7030 [00:05<00:00, 1283.92tile/s]


  Saved npy  : stitched_seg.npy
  Saved png  : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0_segmentation.png

  Class distribution:
    Tumour               946,694 px  (0.3%)  
    Stroma               169,229,982 px  (52.2%)  ████████████████████
    Inflammatory         3,579,466 px  (1.1%)  
    Necrosis             99,536,108 px  (30.7%)  ████████████
    Others               50,811,960 px  (15.7%)  ██████

  GeoJSON extraction
  Canvas   : 49062x25086 px  origin=(3265,9731)
    Tumour               946,694 px  (0.3%)
    Stroma               169,229,982 px  (52.2%)
    Inflammatory         3,579,466 px  (1.1%)
    Necrosis             99,536,108 px  (30.7%)
    Others               50,811,960 px  (15.7%)


Writing GeoJSON:  20%|██████████████████████████████████████                                                                                                                                                        | 1/5 [00:04<00:16,  4.00s/class]

  Tumour                 257 polygons  ~0.53 MB


Writing GeoJSON:  40%|████████████████████████████████████████████████████████████████████████████                                                                                                                  | 2/5 [00:10<00:17,  5.73s/class]

  Stroma               5,714 polygons  ~22.06 MB


Writing GeoJSON:  60%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████                                                                            | 3/5 [00:15<00:10,  5.01s/class]

  Inflammatory         1,927 polygons  ~3.37 MB


Writing GeoJSON:  80%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████                                      | 4/5 [00:20<00:05,  5.13s/class]

  Necrosis             3,741 polygons  ~13.17 MB


Writing GeoJSON: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:26<00:00,  5.37s/class]

  Others               5,681 polygons  ~17.69 MB



  Combined : 17,320 polygons  51.21 MB  → segmentation_all_classes.geojson


Cleaning temporary files: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7031/7031 [00:01<00:00, 5464.62file/s]


  Cleaned up : 7,031 .npy files from segmentation/

  Done.
  Colour PNG   : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0_segmentation.png
  Combined GJ  : segmentation_all_classes.geojson
  .npy removed : 7,031
  ✓  [stitching] done in 124.9s

  ▶  [tumor_roi_overlay] starting …

  Tumor ROI overlay
  Slide      : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  Manifest   : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/segmentation/manifest.csv
  ROI size   : 200 µm  →  800.0 px  (mpp=0.25)
  Output     : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/spatial_feature_results/tumor_roi_overlay


Building ROI boxes: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:00<00:00, 313.43cluster/s]


  Wrote: tumor_roi_boxes.csv  (5 boxes)
  Generated 5 ROI box(es) across 3 tumor clump(s).


Saving pseudo-thumbnail overlay:   0%|                                                                                                                                                                                      | 0/1 [00:01<?, ?image/s]


  Wrote: tumor_roi_boxes_pseudo_thumbnail.png


Saving WSI thumbnail overlay:   0%|                                                                                                                                                                                         | 0/1 [00:02<?, ?image/s]


  Wrote: tumor_roi_boxes_wsi_thumbnail.png
  (WSI metadata mpp-x = 0.5054 µm/px, for reference)

  Done.
  ✓  [tumor_roi_overlay] done in 3.5s

  ▶  [cluster_tils_tsr_score] starting …

  Cluster TSR / sTILs scoring
  Slide      : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  ROI boxes  : tumor_roi_boxes.csv
  GeoJSON    : segmentation_all_classes.geojson
  sTILs denom: salgado
  mpp        : 0.25
  Output     : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/spatial_feature_results/cluster_tils_tsr_score


Dissolving ROI boxes into cluster polygons: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 217.09cluster/s]


  Wrote: cluster_scoring_polygons.geojson


Scoring clusters (TSR / sTILs): 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 21.60cluster/s]


  Wrote: tils_tsr_by_cluster.csv
  Wrote: tils_tsr_wsi_summary.csv


Rendering tissue overlay:  80%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊                                    | 4/5 [00:01<00:00,  2.28class/s]
                                                                                                                                                                                                                                                     /cluster/home/srivash/ViP-SegD/VipSegD_v1/cluster_tils_tsr_scoring.py:1037: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.
  fig.tight_layout()
Saving cluster overlay:   0%|                                                                                                                                                                                               | 0/1 [00:02<?, ?image/s]


  Wrote: cluster_tils_tsr_overlay.png

  === WSI scoring summary ===
  Clusters input / scored : 3 / 3
  ROI boxes  input / scored: 5 / 5
  WSI TSR                 : 4/96  (stroma-high)
  WSI sTILs (Salgado)    : 2.49%
  WSI sTILs (selected)   : 2.49%  [salgado]
  Total cluster area      : 1.7425 mm²
  Tissue fraction         : 0.576

  Done.
  ✓  [cluster_tils_tsr_score] done in 11.6s

  ▶  [immune_proximity] starting …

  Immune proximity features
  Slide    : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  Clusters : cluster_scoring_polygons.geojson
  Output   : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/spatial_feature_results/immune_proximity
  Clusters:             3
  Segmentation regions: 17320


Computing immune proximity: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 16.75cluster/s]

  Wrote: immune_proximity_by_cluster.csv
  Wrote: immune_proximity_wsi_summary.csv


  Wrote: immune_proximity_plot.png

  === Immune proximity summary ===
  Total TIL area:          0.020 mm²
  TIL fraction of cluster: 0.011
  Intratumoral fraction:   0.001
  Median extratumoral dist:50.4 µm
  Dominant phenotype:      peritumoral

  Done.
  ✓  [immune_proximity] done in 10.6s

  ▶  [necrosis_proximity] starting …

  Necrosis proximity features
  Slide    : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  Clusters : cluster_scoring_polygons.geojson
  Output   : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/spatial_feature_results/necrosis_proximity
  Clusters:             3
  Segmentation regions: 17320


Computing necrosis proximity: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00,  9.56cluster/s]


  Wrote: necrosis_proximity_by_cluster.csv
  Wrote: necrosis_proximity_wsi_summary.csv
  Wrote: necrosis_distance_figure.png
  Wrote: necrosis_tissue_context_figure.png

  === Necrosis proximity summary ===
  Total necrosis area:          0.016 mm²
  Necrosis fraction of cluster: 0.009
  Intratumoral fraction:        0.033
  Immune coupling index:        0.034
  Dominant phenotype:           peritumoural

  Done.
  ✓  [necrosis_proximity] done in 9.7s

  ▶  [tumor_morphology] starting …

  Tumour morphology features
  Slide              : TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0
  min_island_area_um2: 1000
  Clusters           : cluster_scoring_polygons.geojson
  Output             : vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/spatial_feature_results/tumor_morphology
  Loaded 257 tumour polygons from 17,320 GeoJSON features.


Computing tumour morphology: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 58.86cluster/s]

  Wrote: tumor_core_features_by_cluster.csv
  Wrote: tumor_core_wsi_summary.csv
  Wrote: tumor_area_by_cluster.png


  Wrote: tumor_fraction_by_cluster.png
  Wrote: tumor_islands_by_cluster.png
  Wrote: tumor_fragmentation_by_cluster.png
  Wrote: tumor_fragmentation_landscape.png

  === Tumour morphology summary ===
  Clusters scored:         3
  Total tumour area:       0.033 mm²
  Total islands (filtered):6
  Mean fragmentation index:0.299
  Mean compactness:        0.355

  Done.
  ✓  [tumor_morphology] done in 5.3s

############################################################
  ✓  Pipeline complete for TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0  (331.1s total)
############################################################

  Stage                            Status         Time
  -------------------------------- ---------- --------
  ✓ tessellation                    done          61.1s
  ✓ segmentation                    done         104.5s
  ✓ stitching                       done         124.9s
  ✓ tumor_roi_overlay               done           3.5s
  ✓ cluster_tils_tsr_score    

## Combined per-cluster feature table

Merge all per-cluster CSVs into one wide feature table — one row per tumour cluster,
all features side by side.  This is the table you would use as input to a
survival model or classifier.


In [7]:
from pathlib import Path
import pandas as pd

slide_name = Path(cfg.WSI_PATH).stem
sf = Path(cfg.OUT_DIR) / slide_name / 'spatial_feature_results'

csv_map = {
    'tils':    sf / 'cluster_tils_tsr_score'  / 'tils_tsr_by_cluster.csv',
    'immune':  sf / 'immune_proximity'         / 'immune_proximity_by_cluster.csv',
    'necrosis':sf / 'necrosis_proximity'       / 'necrosis_proximity_by_cluster.csv',
    'morpho':  sf / 'tumor_morphology'         / 'tumor_core_features_by_cluster.csv',
}

frames = {}
for key, path in csv_map.items():
    if path.exists():
        frames[key] = pd.read_csv(path)
        print(f'  Loaded {key:10s}: {len(frames[key]):>3} rows × {frames[key].shape[1]:>3} cols')
    else:
        print(f'  ✗ {key}: not yet generated')

if frames:
    base = next(iter(frames.values()))[['cluster_id']].copy()
    for key, df in frames.items():
        # Drop columns that exist in other frames to avoid _x / _y collisions
        drop_cols = [c for c in df.columns if c in base.columns and c != 'cluster_id']
        base = base.merge(df.drop(columns=drop_cols), on='cluster_id', how='left')

    print(f'\nCombined table: {base.shape[0]} clusters × {base.shape[1]} features')
    display(base.head())

    # Save to disk
    out_path = sf / 'combined_features_by_cluster.csv'
    base.to_csv(out_path, index=False)
    print(f'\nSaved → {out_path}')


  Loaded tils      :   3 rows ×  56 cols
  Loaded immune    :   3 rows ×  20 cols
  Loaded necrosis  :   3 rows ×  23 cols
  Loaded morpho    :   3 rows ×  38 cols

Combined table: 3 clusters × 124 features


,cluster_id,n_roi_boxes,priority_nonoverlap,cluster_scoring_area_px2,cluster_scoring_area_um2,cluster_scoring_perimeter_px,cluster_scoring_perimeter_um,buffer_um,tumor_area_px2,tumor_area_um2,...,tumor_island_area_median_um2,tumor_island_area_p90_um2,tumor_island_area_max_um2,tumor_island_nnd_mean_um,tumor_island_nnd_median_um,tumor_spread_frac,tumor_convex_hull_fill,min_island_area_um2_used,tumor_valid_morphology,tumor_invalid_reason
0,0,1,2,5.062227e+06,316389.183225,10639.171669,2659.792917,300.0,748.206393,46.762900,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1000.0,True,NaN
1,2,3,1,1.382103e+07,863814.445520,13514.440741,3378.610185,300.0,431430.622335,26964.413896,...,2769.25000,7268.00109,8043.28125,204.101426,166.881554,0.463875,0.199606,1000.0,True,NaN
2,3,1,3,8.996630e+06,562289.364149,10736.794777,2684.198694,300.0,99674.000000,6229.625000,...,5345.03125,5345.03125,5345.03125,NaN,NaN,0.000000,NaN,1000.0,True,NaN



Saved → vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/spatial_feature_results/combined_features_by_cluster.csv


## WSI-level summary

Merge all WSI-level summary rows into one horizontal table.  This gives the
single-row feature vector for the whole slide.

In [8]:
wsi_csv_map = {
    'tils':    sf / 'cluster_tils_tsr_score'  / 'tils_tsr_wsi_summary.csv',
    'immune':  sf / 'immune_proximity'         / 'immune_proximity_wsi_summary.csv',
    'necrosis':sf / 'necrosis_proximity'       / 'necrosis_proximity_wsi_summary.csv',
    'morpho':  sf / 'tumor_morphology'         / 'tumor_core_wsi_summary.csv',
}

wsi_frames = {}
for key, path in wsi_csv_map.items():
    if path.exists():
        wsi_frames[key] = pd.read_csv(path)

if wsi_frames:
    combined_wsi = pd.concat(
        [df.reset_index(drop=True) for df in wsi_frames.values()],
        axis=1
    )
    # Remove duplicate columns (e.g. n_clusters)
    combined_wsi = combined_wsi.loc[:, ~combined_wsi.columns.duplicated()]
    print(f'WSI feature vector: {combined_wsi.shape[1]} features')
    display(combined_wsi.T)

    wsi_out = sf / 'combined_wsi_summary.csv'
    combined_wsi.to_csv(wsi_out, index=False)
    print(f'\nSaved → {wsi_out}')


WSI feature vector: 120 features


,0
tils_denominator_used,salgado
mpp,0.25
n_clusters_input,3
n_clusters_scored,3
n_clusters_filtered_min_roi,0
...,...
wsi_area_weighted_tumor_convex_hull_fill,0.199606
wsi_island_area_mean_um2,4219.081183
wsi_island_area_median_um2,4057.140625
wsi_island_area_p90_um2,7074.18105



Saved → vipsegd_output/TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0/spatial_feature_results/combined_wsi_summary.csv


## Generating the QMD Report File

This cell generates a Quarto Markdown (`.qmd`) report file for the current whole-slide image using values from the pipeline configuration, including the slide path, output directory, author, and notes. The generated `.qmd` file is the report source file and can be rendered from the command line to produce the final HTML report.

To generate the `.html` file after `vispace_report_slide.qmd` is generated in folder run it in bash (this step is only for server users no collab interface support)

```bash
quarto render vispace_report_slide.qmd 


In [10]:
from generate_qmd_file import generate_report
report_qmd = generate_report(
    wsi_path=cfg.WSI_PATH,
    cfg=cfg,
)

print(f"Generated report file: {report_qmd}")

Generated report file: vispace_report_slide.qmd



## Output file inventory

List every file produced by the pipeline, grouped by stage.


In [11]:
import os

slide_name = Path(cfg.WSI_PATH).stem
out_root   = Path(cfg.OUT_DIR) / slide_name

total_size = 0
for dirpath, _, filenames in os.walk(out_root):
    rel = Path(dirpath).relative_to(out_root)
    if filenames:
        print(f'  {rel}/')
        for fname in sorted(filenames):
            fpath = Path(dirpath) / fname
            size  = fpath.stat().st_size
            total_size += size
            size_str = f'{size/1e6:.2f} MB' if size > 1e6 else f'{size/1e3:.1f} KB'
            print(f'    {fname:<55} {size_str:>10}')

print(f'\nTotal output size: {total_size/1e6:.1f} MB')

  segmentation/
    TCGA-S3-AA15-01Z-00-DX1.A2456A4A-E6E8-4429-8F09-B997AA497BB0_segmentation.png   24.66 MB
    manifest.csv                                               1.35 MB
    segmentation_all_classes.geojson                          51.21 MB
    stitched_seg_metadata.json                                  0.7 KB
  spatial_feature_results/
    combined_features_by_cluster.csv                            7.3 KB
    combined_wsi_summary.csv                                    4.7 KB
  spatial_feature_results/cluster_tils_tsr_score/
    cluster_scoring_polygons.geojson                            9.5 KB
    cluster_tils_tsr_overlay.png                               1.07 MB
    tils_tsr_by_cluster.csv                                     3.6 KB
    tils_tsr_wsi_summary.csv                                    1.9 KB
  spatial_feature_results/necrosis_proximity/
    necrosis_distance_figure.png                              124.9 KB
    necrosis_proximity_by_cluster.csv                     